Add enrich policy runner

jbaiera opened the pull request:
Note that this is a PR against the enrich branch and will be backported to enrich-7.x branch.

Adds a skeleton of the execution logic to execute an enrich policy. Validates the source index existence as well as mappings, creates a new enrich index for the policy, reindexes the source index into the new enrich index, and swaps the enrich alias for the policy to the new index.

~WIP: Comments are welcome at this point. The code really needs some testing and will most likely be rebased on to some other PR's as they come in (namely #41003 and #40997), but I'd like to get some feedback before then.~ Good to go for review

martijnvg commented:
The general approach in the PR looks good to me.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,238 @@
+        ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetaData>> mappings = getIndexResponse.mappings();
+        ImmutableOpenMap<String, MappingMetaData> indexMapping = mappings.get(sourceIndexName);
+        // Assuming single mapping
+        MappingMetaData typeMapping = indexMapping.iterator().next().value;
```
maybe add an assert that there should be only a single mapping?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,238 @@
+        logger.debug("Policy [{}]: Validating [{}] source mappings", policyName, sourceIndexPattern);
+        Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndexPattern);
+        @SuppressWarnings("unchecked")
+        Map<String, Object> properties = ((Map<String, Object>) mapping.get("properties"));
```
maybe `Map<?, ?> properties` in order to remove the suppress warning annotation?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyResult.java:
```diff
@@ -0,0 +1,22 @@
+package org.elasticsearch.xpack.enrich;
+
+public class EnrichPolicyResult {
+    boolean completed;
```
`private boolean completed`?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,238 @@
+    private IndexNameExpressionResolver indexNameExpressionResolver;
+    private LongSupplier nowSupplier;
+
+    public EnrichPolicyRunner(ClusterService clusterService, Client client, EnrichStore enrichStore,
```
maybe make this constructor package protected?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,238 @@
+        // TODO: Minimize the mappings here, merge mappings from multiple indices
+        Map<String, Object> mapping = getMappings(getIndexResponse, sourceIndexName);
+        @SuppressWarnings("unchecked")
+        Map<String, Object> properties = ((Map<String, Object>) mapping.get("properties"));
```
same as the other comment about the suppress warning annotation?

martijnvg commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -0,0 +1,141 @@
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.nullValue;
+
+public class EnrichPolicyRunnerTests extends ESSingleNodeTestCase {
```
would be great to test this without starting a node, but I'm not sure how :)

jbaiera replied:
Yeah, this was one of those cases where it felt like mocking out all of the dependencies needed was a bit much. Would be happy to change the test if there are any tools to make that easier, or if we feel that it's important enough that it doesn't spin up a node.

martijnvg replied:
Yes, I think this is the right trade off. Otherwise the code for mocking becomes unmaintainable.

jbaiera commented:
Removed WIP label, should be good for a proper review at this point

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,299 @@
+            .map();
+
+        Map<String, Object> enrichKeyTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match", policy.getEnrichKey())
```
I think we should use `path_match` here (and unmatch) to allow for nested object keys

jbaiera replied:
I'm going to throw out the mapping templates and just use a regular mapping for the keys.

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+
+    private Map<String, Object> resolveEnrichMapping(final EnrichPolicy policy) {
+        // TODO: Modify enrichKeyFieldMapping based on policy type.
+        // Create dynamic mapping templates - one for the enrich key and one for the enrich values
```
Is it possible to make this a bit more readable via a single map with indentions ?

..or a `JsonXContent.contentBuilder().startObject()` and `Strings.toString(templateBuilder)`

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+            .map();
+
+        Map<String, Object> enrichTextValueTemplateDefinition = MapBuilder.<String, Object>newMapBuilder()
+            .put("match_mapping_type", "string")
```
Why do we match on Strings ? (if it is for the 256 limit, we should allow larger for values and smaller for keys... possibly a todo to expose these as soft limits ?)

jbaiera replied:
This is for the case that an enrich value is mapped as a `text` field. Copying a `text` field as-is from a source index does not work since `doc_values` [cannot be enabled for them.](https://www.elastic.co/guide/en/elasticsearch/reference/current/doc-values.html)

jbaiera replied:
The 256 limit simply came from the default keyword mapping. I'm fine with modifying that.

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+        client.execute(ReindexAction.INSTANCE, reindexRequest, new ActionListener<BulkByScrollResponse>() {
+            @Override
+            public void onResponse(BulkByScrollResponse bulkByScrollResponse) {
+                // Do we want to fail the request if there were failures during the reindex process?
```
I am not familiar with this API... does this get called once after the whole re-index is complete, or per scroll response ?

martijnvg replied:
this gets invoked after the whole reindex has been completed.

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyResult.java:
```diff
@@ -0,0 +1,18 @@
+ */
+package org.elasticsearch.xpack.enrich;
+
+public class EnrichPolicyResult {
```
nit: `EnrichPolicyResult` -> `EnrichPolicyRunResult`

martijnvg replied:
maybe `PolicyExecutionResult`?

jakelandis commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -0,0 +1,252 @@
+    }
+
+    public void testRunner() throws Exception {
+        logger.info("Starting Test");
```
Does this end up in the CI output ?

martijnvg replied:
Yes, but is only visible if a test failed.

jbaiera replied:
I'll scrub this - it was more for my own debugging while building out the initial test.

jakelandis commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -0,0 +1,252 @@
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.nullValue;
+
```
The tests here are good for the happy case, but I would like to see more testing around the un-happy cases. Investments now in this test will greatly help future developers to quickly test changes.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+        final String sourceIndexPattern = policy.getIndexPattern();
+        logger.debug("Policy [{}]: Checking source index [{}]", policyName, sourceIndexPattern);
+        GetIndexRequest getIndexRequest = new GetIndexRequest().indices(sourceIndexPattern);
+        client.admin().indices().getIndex(getIndexRequest, new ActionListener<GetIndexResponse>() {
```
new ActionListener<GetIndexResponse> -> new ActionListener<>

jbaiera replied:
To be clear here, do you mean `new ActionListener<GetIndexResponse>()` -> `new ActionListener<>()`? That change would end up with compilation errors

martijnvg replied:
this: `client.admin().indices().getIndex(getIndexRequest, new ActionListener<>() {` does compile here without an error.

jbaiera replied:
I get `Cannot use "<>" with anonymous inner classes` in IDEA, and the `onResponse` override only accepts `Object` as its argument. Maybe I'm doing something wrong. Looking at some random `ActionListener` anonymous implementations and they all declare a type instead of using the diamond operator.

jbaiera replied:
Strange, IDEA seems to be the one with the hangup. Gradle seems to build this just fine.

martijnvg replied:
Odd. It compiled for me in both IDEA and gradle. Update IDEA?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+        CreateIndexRequest createEnrichIndexRequest = new CreateIndexRequest(enrichIndexName, enrichIndexSettings);
+        createEnrichIndexRequest.mapping("_doc", enrichMapping);
+        logger.debug("Policy [{}]: Creating new enrich index [{}]", policyName, enrichIndexName);
+        client.admin().indices().create(createEnrichIndexRequest, new ActionListener<CreateIndexResponse>() {
```
new ActionListener<CreateIndexResponse> -> new ActionListener<>

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+            .setDestIndex(destinationIndexName)
+            .setSourceIndices(policy.getIndexPattern());
+        reindexRequest.getSearchRequest().source(searchSourceBuilder);
+        client.execute(ReindexAction.INSTANCE, reindexRequest, new ActionListener<BulkByScrollResponse>() {
```
new ActionListener<BulkByScrollResponse> -> new ActionListener<>

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+    private void refreshEnrichIndex(final String destinationIndexName, final String policyName,
+                                    final ActionListener<EnrichPolicyResult> listener) {
+        logger.debug("Policy [{}]: Refreshing newly created enrich index [{}]", policyName, destinationIndexName);
+        client.admin().indices().refresh(new RefreshRequest(destinationIndexName), new ActionListener<RefreshResponse>() {
```
new ActionListener<RefreshResponse> -> new ActionListener<>

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+            aliasToggleRequest.addAliasAction(IndicesAliasesRequest.AliasActions.remove().indices(indices).alias(enrichIndexBase));
+        }
+        aliasToggleRequest.addAliasAction(IndicesAliasesRequest.AliasActions.add().index(destinationIndexName).alias(enrichIndexBase));
+        client.admin().indices().aliases(aliasToggleRequest, new ActionListener<AcknowledgedResponse>() {
```
new ActionListener<AcknowledgedResponse> -> new ActionListener<>

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+
+    public void runPolicy(final String policyId, final ActionListener<EnrichPolicyResult> listener) {
+        // Look up policy in policy store and execute it
+        runPolicy(policyId, enrichStore.getPolicy(policyId), listener);
```
I think we need to handle to case that the EnrichStore returns null here and then throw an appropriate exception.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,299 @@
+                        "Enrich policy execution for [{}] failed. Could not locate enrich key field [{}] on mapping for index [{}]",
+                        policyName, policy.getEnrichKey(), sourceIndex));
+            }
+            for (String enrichField : policy.getEnrichValues()) {
```
I lean towards failing if a decorate field is missing in the mapping. I think this can lead to unexpected behaviour at enrich time? (why does this document not have all the enriched fields?)

jakelandis replied:
Even if the mapping exists in the source index, the document we are using to enrich may not have that field and will need handle partial decorations in the processor. 

If the decision to only allow full decorations, then the check here makes sense... but will need to changed slightly such that all enrich fields exist across the set of the source indexes (as opposed to exist in every source index as it is here). 

I am in favor of removing this check and implement `ignore_missing` in the processor to allow partial decorations.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+            .put("enrich_value_template", enrichValueTemplateDefinition)
+            .map();
+
+        List<Map<String, Object>> templates = new ArrayList<>();
```
I should have mentioned this yesterday. I'm currently developing the enrich processor and we no longer need define a mapping for each enrich value field separately. I'm working on a custom meta field type, that does something similar that the `SourceFieldMapper` is doing, but instead stores the source of a document in binary doc values. 

The advantage is that at enrich time we only need to do a single doc values lookup instead of multiple. This going to be more efficient in case that there are multiple enrich value fields, which I think will be the case.

The other consequence is that enrich value fields shouldn't be defined in the mapping of an enrich index. The mapping will look something like:

```
{
   "_source": {"enabled": false}
   "_enrich_source": {"enabled: true}
   "properties": {
       "enrich_key": {
          "type": "keyword",
          "doc_values": false
       }
   }
}
```

jbaiera replied:
This makes things so much more simple, thanks.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+    private Map<String, Object> resolveEnrichMapping(final EnrichPolicy policy) {
+        // TODO: Modify enrichKeyFieldMapping based on policy type.
+        // Create dynamic mapping templates - one for the enrich key and one for the enrich values
+        Map<String, Object> enrichKeyFieldMapping = MapBuilder.<String, Object>newMapBuilder()
```
I don't think we need template fields for the enrich key field? We know the name of the field from the policy and the policy type determines the field type. In case of exact match that is `keyword`. So I think we should just use a concrete mapping here.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,238 @@
+import org.elasticsearch.index.reindex.ReindexRequest;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+
+public class EnrichPolicyRunner {
```
As soon as the get index request is executed this is asynchronous. The ActionListener call back is invoked by another thread (likely network thread) and then delegates to the next step where another api is invoked. So if this class is a runnable then that doesn't buy us much?

jakelandis replied:
The GetIndexRequest defaults to `ThreadPool.Names.SAME` which is executes on the current thread (likely the network thread) See [comment](https://github.com/elastic/elasticsearch/pull/41088#discussion_r275960708) above. Expressing this class as a `Runnable` and introducing another class to fork then run it(`EnrichPolicyExecutor`) is just a suggestion to help indicate that this whole thing is ran in async. Also has the benefits that consumers either via the REST API (e.g. _execute) or the background scheduler have a common place (`EnrichPolicyExecutor`) to execute the runner without worrying about threading themselves.

jakelandis commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,298 @@
+                        "Enrich policy execution for [{}] failed. Could not read mapping for source [{}] included by pattern [{}]",
+                        policyName, sourceIndex, policy.getIndexPattern()));
+            }
+            if (properties.containsKey(policy.getEnrichKey()) == false) {
```
we should probably allow `_id` as the enrich key

jbaiera replied:
Agreed, though I think it's probably a good idea to put that into a separate PR. I don't think we should bog this one down with too many initial features.

jbaiera commented:
I made the runner object a Runnable and added a basic EnrichPolicyExecutor implementation. I figured the management threadpool would be fine to execute these on. I also simplified the enrich mapping to just define the key, and rely on the eventual meta mapping type to be added later.

martijnvg commented on x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java:
```diff
@@ -0,0 +1,231 @@
+        GetIndexResponse enrichIndex = client().admin().indices()
+            .getIndex(new GetIndexRequest().indices(".enrich-test1")).get();
+        assertThat(enrichIndex.getIndices().length, equalTo(1));
+        assertThat(enrichIndex.getIndices()[0], equalTo(".enrich-test1-" + createTime));
```
maybe also assert that the enrich key field has been defined with the correct attributes set, _source is disabled, dynamic field mapping is set to false and auto expand replicas are enabled?

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,259 @@
+            builder.startObject()
+                .startObject(MapperService.SINGLE_MAPPING_NAME)
+                    .startObject("_source")
+                        .field("enabled", false)
```
For now _source needs to be enabled. #41532 adds the enrich processor, but uses the _source for now. A follow-up PR will introduce a special field type that stores the enrich document's decorate fields in a special doc values field. This is going to be added after the demo.

martijnvg commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,259 @@
+        ReindexRequest reindexRequest = new ReindexRequest()
+            .setDestIndex(destinationIndexName)
+            .setSourceIndices(policy.getIndexPattern());
+        reindexRequest.getSearchRequest().source(searchSourceBuilder);
```
The xcontent type in the enrich index should be smile. So we will need to add the following statement here:

```
reindexRequest.getDestination().source(new BytesArray(), XContentType.SMILE)
```

The `destination` field inside `ReindexRequest` is now actually used, but used as a prototype for specific properties. The source isn't used and can therefore be empty in the above statement. (See `TransportReindexAction#buildRequest(...)` method)

jbaiera commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,259 @@
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.fetchSource(retainFields.toArray(new String[0]), new String[0]);
+        if (policy.getQuery() != null) {
+            searchSourceBuilder.query(QueryBuilders.wrapperQuery(policy.getQuery().getQuery()));
```
I went with the wrapper query since the EnrichPolicy.QuerySource returns the raw query json as a byte sequence. Is there a more appropriate way to convert a raw query body into a query builder?

martijnvg replied:
I see, actually that does make sense to me :)

jbaiera commented on x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunner.java:
```diff
@@ -0,0 +1,259 @@
+                listener.onFailure(e);
+            }
+        });
+    }
```
These should definitely be subject to some kind of background cleanup task.
